### DISCUSSION TITLE
How to Log Server Response StatusCode

### DISCUSSION THREAD

**Discussion by simon5454** [vercel/next.js#42910] [Help] [answered]
I am looking for a way to log the outgoing statusCode of the response. Something like morgan with Express.

In ServerResponse there is a listener called finish or a statusCode that should be possible to use. But, it always returns a 200 status even when the correct returned status is 404 for example.

I did try something like this : 

//middleware.ts
export function middleware(request: NextRequest) {
	if (request.nextUrl.pathname.startsWith('/api')) {
		// This logic is only applied to /api
		const response = NextResponse.next();
		consoleLogger.log(response);
		return response;
	}
}

**Comment by icyJoseph** [comment, marked as answer]
This expectation is because you are thinking of middleware, and I don't blame you for that, as passing along the request, but rather, it is a nuanced process. 

When you return `NextResponse.next()` it is not running the API call and then giving you back the result, rather, you tell the middleware engine to forward this request to its next consumer, that one in turn does the transaction it has to do.

Middleware in this case is `in-` only, there's no `out-` (as in back to the viewer) intervention from the middleware.

I am not really sure there's a one stop place to do this, other than wrapping every handler with a logging higher-order function, or creating a custom server that reports the events. 

It becomes even more tricky when considering how middleware and API routes run in some environments, like Vercel, where they are run as serverless lambdas, or at the edge.

**Reply by simon5454** [reply]
Ok, I guess you mean something like this (https://github.com/vercel/next.js/discussions/30681#discussioncomment-1566354)?

I was hoping for a global solution but I can work with that (just more manual work).

**Reply by icyJoseph** [reply]
yeah 😓 it'd have hoped for such a thing too, but it is not there yet, for example it is a bit annoying to lay error handlers to every single API route
